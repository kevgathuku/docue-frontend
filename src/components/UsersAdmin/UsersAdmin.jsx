import React from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import RoleActions from '../../actions/RoleActions';
import RoleStore from '../../stores/RoleStore';

import {
  fetchUsers,
  initiateUpdateProfile,
} from '../../actions/actionCreators';

export class UsersAdmin extends React.Component {
  static propTypes = {
    fetchUsers: PropTypes.func.isRequired,
    updateProfile: PropTypes.func.isRequired,
    users: PropTypes.array,
  };

  constructor(props) {
    super(props);

    this.state = {
      token: localStorage.getItem('user'),
      selectedRole: null,
      roles: [],
      access: {
        viewer: 'Public Documents',
        staff: 'Staff and Public Documents',
        admin: 'All Documents',
      },
      options: [],
    };
  }

  componentDidMount() {
    RoleActions.getRoles(this.state.token);
    this.props.fetchUsers(this.state.token);
    RoleStore.addChangeListener(this.handleRolesResult);
  }

  componentWillUnmount() {
    RoleStore.removeChangeListener(this.handleRolesResult);
  }

  handleRolesResult = () => {
    let roles = RoleStore.getRoles();
    this.setState({ roles: roles });
  };

  // Prepend the user object to the function arguments through bind
  handleSelectChange = (user, val) => {
    // Update the user's Role
    // Don't update if the already existing role is the one chosen
    if (user.role._id !== val._id) {
      user.role = val;

      this.props.updateProfile(user._id, user, this.state.token);
    }
  };

  render() {
    const renderUser = (user) => {
      const access = user.role.title;
      const description = this.state.access[access];
      return (
        <tr key={user._id}>
          <td>{`${user.name.first} ${user.name.last}`}</td>
          <td>{user.email}</td>
          <td>
            <Select
              getOptionLabel={(option) => {
                return option.title;
              }}
              getOptionValue={(option) => {
                return option._id;
              }}
              styles={{ control: (base) => ({ ...base, maxHeight: '50px' }) }}
              name="role"
              options={this.state.roles}
              onChange={this.handleSelectChange.bind(null, user)}
              placeholder="Select Role"
              value={user.role}
              isSearchable={false}
            />
          </td>
          <td>{description}</td>
        </tr>
      );
    };
    return (
      <div className="container">
        <div className="card-panel">
          <h2 className="header center-align">Manage Users</h2>
          <div className="row">
            <div className="col s10 offset-s1 center-align">
              <table className="centered">
                <thead>
                  <tr>
                    <th data-field="id">Name</th>
                    <th data-field="name">Email</th>
                    <th data-field="role">Role</th>
                    <th data-field="role">Access</th>
                  </tr>
                </thead>
                <tbody>{(this.props.users || []).map(renderUser)}</tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    users: state.users,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateProfile: (userId, updatedUserObject, token) => {
      dispatch(initiateUpdateProfile(userId, updatedUserObject, token));
    },
    fetchUsers: (token) => {
      dispatch(fetchUsers(token));
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UsersAdmin);
