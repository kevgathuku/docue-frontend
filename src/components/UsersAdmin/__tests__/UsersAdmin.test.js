import React from 'react';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import RoleStore from '../../../stores/RoleStore';
import RoleActions from '../../../actions/RoleActions';
import { UsersAdmin } from '../UsersAdmin';

describe('UsersAdmin', function() {
  let props;
  beforeEach(() => {
    props = { fetchUsers: jest.fn(), updateProfile: jest.fn() };
  });

  describe('Component Rendering', function() {
    it('renders correctly given no users', function() {
      const wrapper = shallow(<UsersAdmin {...props} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders correctly when users are provided', function() {
      const users = [
        {
          _id: 'somehting',
          email: 'some@ht.ing',
          name: { first: 'efwfeww', last: '' },
          role: { title: 'fefefe', _id: 'roleID' },
        },
      ];
      const localProps = Object.assign({}, props, {
        users,
      });
      const wrapper = shallow(<UsersAdmin {...localProps} />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('Class Functions', function() {
    let usersAdmin;

    beforeEach(function() {
      sinon.stub(RoleActions, 'getRoles').returns(true);
      usersAdmin = shallow(<UsersAdmin {...props} />);
    });

    afterEach(function() {
      RoleActions.getRoles.restore();
    });

    describe('handleSelectChange', function() {
      it('calls updateProfile when the user role and selected role are different', () => {
        const state = { token: 'bar' };
        usersAdmin.setState(state);

        let user = { _id: 1, role: { _id: 22, title: 'Default' } };
        let val = { _id: 1, title: 'staff' };
        const updatedUser = Object.assign({}, user, { role: val });
        const instance = usersAdmin.instance();
        instance.handleSelectChange(user, val);

        expect(props.updateProfile).toHaveBeenCalledWith(
          user._id,
          updatedUser,
          state.token
        );
      });

      it('does not call updateProfile when the user role and selected role are different', () => {
        let role = { _id: 1, title: 'staff' };
        let user = { _id: 1, role };
        const instance = usersAdmin.instance();
        instance.handleSelectChange(user, role);

        expect(props.updateProfile).not.toHaveBeenCalled();
      });
    });

    describe('handleRolesResult', function() {
      it('should correctly update the state', function() {
        let val = [
          {
            _id: 1,
          },
        ];
        RoleStore.setRoles(val);
        expect(RoleStore.getRoles()).toEqual(val);
        expect(usersAdmin.state().roles).toBe(val);
      });
    });
  });
});
