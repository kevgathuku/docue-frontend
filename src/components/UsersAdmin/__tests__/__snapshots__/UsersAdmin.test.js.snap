// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersAdmin Component Rendering renders correctly given no users 1`] = `
<div
  className="container"
>
  <div
    className="card-panel"
  >
    <h2
      className="header center-align"
    >
      Manage Users
    </h2>
    <div
      className="row"
    >
      <div
        className="col s10 offset-s1 center-align"
      >
        <table
          className="centered"
        >
          <thead>
            <tr>
              <th
                data-field="id"
              >
                Name
              </th>
              <th
                data-field="name"
              >
                Email
              </th>
              <th
                data-field="role"
              >
                Role
              </th>
              <th
                data-field="role"
              >
                Access
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`UsersAdmin Component Rendering renders correctly when users are provided 1`] = `
<div
  className="container"
>
  <div
    className="card-panel"
  >
    <h2
      className="header center-align"
    >
      Manage Users
    </h2>
    <div
      className="row"
    >
      <div
        className="col s10 offset-s1 center-align"
      >
        <table
          className="centered"
        >
          <thead>
            <tr>
              <th
                data-field="id"
              >
                Name
              </th>
              <th
                data-field="name"
              >
                Email
              </th>
              <th
                data-field="role"
              >
                Role
              </th>
              <th
                data-field="role"
              >
                Access
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              key="somehting"
            >
              <td>
                efwfeww 
              </td>
              <td>
                some@ht.ing
              </td>
              <td>
                <StateManager
                  defaultInputValue=""
                  defaultMenuIsOpen={false}
                  defaultValue={null}
                  getOptionLabel={[Function]}
                  getOptionValue={[Function]}
                  isSearchable={false}
                  name="role"
                  onChange={[Function]}
                  options={Array []}
                  placeholder="Select Role"
                  styles={
                    Object {
                      "control": [Function],
                    }
                  }
                  value={
                    Object {
                      "_id": "roleID",
                      "title": "fefefe",
                    }
                  }
                />
              </td>
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
